lib/tests/auth/cognito.test.ts

```diff
@@ -0,0 +1,313 @@
+/**
+ * @jest-environment node
+ */
+
+import Redis from "ioredis-mock";
+import { mockClient } from "aws-sdk-client-mock";
+import {
+  CognitoIdentityProviderClient,
+  AdminInitiateAuthCommand,
+  AdminInitiateAuthResponse,
+} from "@aws-sdk/client-cognito-identity-provider";
+import { prismaMock } from "@jestUtils";
+import {
+  initiateSignIn,
+  begin2FAAuthentication,
+  requestNew2FAVerificationCode,
+  validate2FAVerificationCode,
+  Validate2FAVerificationCodeResultStatus,
+} from "@lib/auth/cognito";
+import { Base } from "__utils__/permissions";
+import { generateVerificationCode, sendVerificationCode } from "@lib/auth/2fa";
+import { registerFailed2FAAttempt, clear2FALockout } from "@lib/auth/2faLockout";
+
+const redis = new Redis();
+jest.mock("@lib/integration/redisConnector", () => ({
+  getRedisInstance: jest.fn(() => redis),
+}));
+
+const cognitoMock = mockClient(CognitoIdentityProviderClient);
+
+const mockSendEmail = {
+  sendEmail: jest.fn(() => Promise.resolve()),
+};
+jest.mock("notifications-node-client", () => ({
+  NotifyClient: jest.fn(() => mockSendEmail),
```

Comment:
```suggestion
  NotifyClient: jest.fn(),
```
There's no need to go a level deeper to mock the `sendEmail` function unless you wanted to test that `sendEmail` has been called.  Jest mock will stub out the initialization class and all sub functions.

Reply:
That is right! I think I was planning on checking if it was called but then decided to use `expect(mockSendVerificationCode).toHaveBeenCalled();` instead of `sendEmail`